File: Paid_ChatGPT/ordenTopologico/orden_topologico.js
function esOrdenTopologico(grafo, orden) {
    // grafo: objeto donde cada clave es un nodo y su valor es lista de adyacentes
    // orden: arreglo con el supuesto orden topológico

    const posicion = new Map();

    for (let i = 0; i < orden.length; i++) {
        posicion.set(orden[i], i);
    }

    for (const nodo in grafo) {
        if (!posicion.has(nodo)) {
            return false;
        }

        for (const vecino of grafo[nodo]) {
            if (!posicion.has(vecino)) {
                return false;
            }

            if (posicion.get(nodo) > posicion.get(vecino)) {
                return false;
            }
        }
    }

    return true;
}

// Ejemplo de uso
const grafo = {
    A: ["C"],
    B: ["C", "D"],
    C: ["E"],
    D: ["F"],
    E: [],
    F: []
};

const orden1 = ["A", "B", "C", "D", "E", "F"];
const orden2 = ["C", "A", "B", "D", "E", "F"];

console.log(esOrdenTopologico(grafo, orden1)); // true
console.log(esOrdenTopologico(grafo, orden2)); // false

File: Paid_ChatGPT/ordenTopologico/orden_topologico.rs
use std::collections::{HashMap, HashSet};

fn es_orden_topologico(
    grafo: &HashMap<&str, Vec<&str>>,
    orden: &[&str],
) -> bool {
    let mut nodos: HashSet<&str> = HashSet::new();

    for (u, vecinos) in grafo {
        nodos.insert(u);

        for v in vecinos {
            nodos.insert(v);
        }
    }

    if orden.iter().copied().collect::<HashSet<_>>() != nodos {
        return false;
    }

    let mut posicion: HashMap<&str, usize> = HashMap::new();

    for (i, nodo) in orden.iter().enumerate() {
        posicion.insert(nodo, i);
    }

    for (u, vecinos) in grafo {
        for v in vecinos {
            if posicion[u] >= posicion[v] {
                return false;
            }
        }
    }

    true
}

fn main() {
    let mut grafo: HashMap<&str, Vec<&str>> = HashMap::new();

    grafo.insert("A", vec!["C"]);
    grafo.insert("B", vec!["C", "D"]);
    grafo.insert("C", vec!["E"]);
    grafo.insert("D", vec!["E"]);
    grafo.insert("E", vec![]);

    let orden1 = vec!["A", "B", "C", "D", "E"];
    let orden2 = vec!["C", "A", "B", "D", "E"];

    println!("{}", es_orden_topologico(&grafo, &orden1)); // true
    println!("{}", es_orden_topologico(&grafo, &orden2)); // false
}